### PR TITLE
validation: make environment property mandatory in serial steps

### PR DESCRIPTION
**Pull request by giuseppe-steduto**
Closes reanahub/reana-client#679.

### CODE REVIEW

**Review by mdonadoni — COMMENTED** [review]
This seems to break some non-serial workflows, such as the Snakemake version of helloworld. Can you reproduce this?

```console
$ reana-client run -f reana-snakemake.yaml
==> Creating a workflow...
Job stats:
job           count    min threads    max threads
----------  -------  -------------  -------------
all               1              1              1
helloworld        1              1              1
total             2              1              1

[WARNING] Job stats:
job           count    min threads    max threads
----------  -------  -------------  -------------
all               1              1              1
helloworld        1              1              1
total             2              1              1

==> Verifying REANA specification file... .../reana-demo-helloworld/reana-snakemake.yaml
[ERROR] Invalid REANA specification: None is not of type 'string'
==> ERROR: Cannot create workflow : 
None is not of type 'string'

Failed validating 'type' in schema['properties']['workflow']['properties']['specification']['properties']['steps']['items']['properties']['kubernetes_memory_limit']:
    {'id': '/properties/workflow/properties/specification/properties/steps/properties/kubernetes_memory_limit',
     'title': 'Memory limit for the step container (e.g. 256Mi).',
     'type': 'string'}

On instance['workflow']['specification']['steps'][0]['kubernetes_memory_limit']:
    None
```

Also note that the CWL version has a very different `workflow.specification` (even though it still works fine):
```yaml
"workflow": {
  "file": "workflow/cwl/helloworld.cwl",
  "specification": {
    "$graph": [
      {
        "class": "Workflow",
        "id": "#main",
        "inputs": [
          {
            "id": "#main/helloworld",
            "type": "File"
...
```

Maybe it's possible to check the schema of `workflow.specification` only if `workflow.type` is serial?

**Comment by giuseppe-steduto** [comment]
> This seems to break some non-serial workflows, such as the Snakemake version of helloworld

Apparently this happens because the Snakemake specification is loaded using `None` as a value for all the keys that were not explicitely specified in the Snakefile. Thus a solution could be to allow for both strings and null values for this and other properties.

> Maybe it's possible to check the schema of workflow.specification only if workflow.type is serial?

AFAIK in JSON Schema draft-04 (the one we are using), one can't express conditional requirements directly. JSON Schema draft-07 introduced the `if`, `then`, and `else` structure, so it would be possible to do something like this:
```json
{
  ...
  "workflow": {
    ...
    "properties": {
      "type": {
        "type": "string",
        "enum": [
          "cwl",
          "serial",
          "yadage",
          "snakemake"
        ]
      }
    },
    "if": {
      "properties": {
        "type": {
          "const": "serial"
        }
      }
    },
    "then": {
      "properties": {
        "specification": {
          "properties": {
            "steps": {
              "items": {
                "required": ["environment"]
              }
            }
          }
        }
      }
    },
    "else": {
      "properties": {
        "specification": {
          "id": ...
        }
      }
    }
  }
}
```

But upgrading to draft-07 might not be worth it, especially if it's only for this reason (draft-06 introduced [a few big backwards-incompatible changes](https://json-schema.org/draft-06/json-schema-release-notes#backwards-incompatible-changes)).

Another valid option at this point would be to add only the `"required": ["environment"]"` clause and forget about the rest of the `"specification"` schema. In this way the effect would be:
- Serial workflows: make `environment` mandatory in the steps, as it should be
- Snakemake workflows: make `environment` mandatory in the steps, as it should be (since the steps are loaded in a very similar way to the serial workflows)
- CWL workflows: the steps are inside `specification.$graph.steps`, so the `required` clause does not apply
- Yadage workflows: the steps are inside `specification.stages.scheduler...`, so the `required` clause does not apply

In this way, also, all the specifications covered by `reana-dev run-example -w cwl -w snakemake -w serial -w yadage -c DEMO -b kubernetes --submit-only ` seem to pass the validation stage successfully.


To recap:
Solution :one:: allow for string an null values on `kubernetes_memory_limit` and other properties.
Solution :two:: upgrade to JSON schema draft-07 and use the `if`/`then` construct.
Solution :three:: only keep the `"required": ["environment"]"` clause (I would personally go for this one).

What do you think?

**Comment by tiborsimko** [comment]
> What do you think?

Thanks for the nice analysis. IMO the soluttion 3 will do for the problem at hand, but we should be ready to improve the behaviour during sandboxing sprint.

For example, I've heard a wish from Snakemake users who typically use the same container for all the steps that they did not really expect having to specify the `container:` clause for each rule in the Snakefile, because it's not really required when running Snakemake "locally". The users wished to specify a "base container" for the workflow that would be used for the execution of all the rules unless overridden. The current PR is actually very good for catching these must-have-container-clause situations early, but there are chances we may amend this later to allow specifying a base container for the workflow aiming at having 1:1 correspondence between Snakefile-outside-REANA and Snakefile-inside-REANA for these kind of usages. If we do that, then should we revisit the environment validation anyway.

Ideally, it would be nice to have different schemas for different workflow engines, or at least conditionals, so that we can express truly all the constraints, and cover other fields, not only environments.  But this would be a bigger change that could probably go together with the sandboxing sprint...

File: reana_commons/validation/schemas/reana_analysis_schema.json [hunk 1]
```diff
@@ -89,7 +89,10 @@
         },
         "resources": {
           "$id": "/properties/workflow/properties/resources",
-          "type": "object",
+          "type": [
+            "object",
+            "null"
```
Comment (mdonadoni) [inline reana_commons/validation/schemas/reana_analysis_schema.json:94]: In which cases `resources` can be null?

File: reana_commons/validation/schemas/reana_analysis_schema.json [hunk 2]
```diff
@@ -200,5 +204,127 @@
...
+                      "description": "List of commands to be run in the step."
+                    },
+                    "compute_backend": {
+                      "type": "string",
+                        "enum": [
+                          "kubernetes",
+                          "htcondor",
+                          "htcondorcern",
+                          "slurm",
+                          "slurmcern"
```
Comment (mdonadoni) [inline reana_commons/validation/schemas/reana_analysis_schema.json:255]: I do not think `slurm` and `htcondor` are allowed here.  Did you find any places where they are used?

File: tests/test_validation.py [hunk 3]
```diff
@@ -57,7 +66,12 @@ def test_warnings_reana_yaml(
     warnings = validate_reana_yaml(reana_yaml)
     assert set(expected_warnings.keys()) == set(warnings.keys())
     for key, value in expected_warnings.items():
-        assert set(value) == set(warnings[key])
+        if isinstance(value, list):
+            assert len(value) == len(warnings[key])
+            for warning_value in value:
+                assert warning_value in warnings[key]
+        else:
+            assert operator.eq(value, warnings[key])
```
Comment (mdonadoni) [inline tests/test_validation.py:74]: Just as a curiosity, why we call `operator.eq` here and not `==`?

File: reana_commons/validation/schemas/reana_analysis_schema.json (hunk 1 above)
Comment (giuseppe-steduto) [reply reana_commons/validation/schemas/reana_analysis_schema.json:94]: Never - restored it to just allowing objects

File: reana_commons/validation/schemas/reana_analysis_schema.json (hunk 2 above)
Comment (giuseppe-steduto) [reply reana_commons/validation/schemas/reana_analysis_schema.json:255]: I put them there to be less restrictive towards other REANA deployments that might want to integrate HTCondor or Slurm, but I don't know how the integration with those computational backends would work if external to CERN. Do you think it is better to remove them (and also avoid that CERN users mistakenly write `"htcondor"` instead of `"htcondorcern"`)?

File: tests/test_validation.py (hunk 3 above)
Comment (giuseppe-steduto) [reply tests/test_validation.py:74]: Relic of a previous version of the test in which I used `map` and passed `operator.eq` as a function. I've changed it to `==` to be more pythonic :wink:

**Review by mdonadoni — APPROVED** [review]
Looks good to me!

The only thing that needs to be changed is the new alpha version.

File: reana_commons/version.py [hunk 4]
```diff
@@ -14,4 +14,4 @@
 
 from __future__ import absolute_import, print_function
 
-__version__ = "0.9.4a2"
+__version__ = "0.9.4a4"
```
Comment (mdonadoni) [inline reana_commons/version.py:17]: Let's regenerate the release commit so that the new version is 0.9.4a3 and not 0.9.4a4

File: reana_commons/validation/schemas/reana_analysis_schema.json (hunk 2 above)
Comment (mdonadoni) [reply reana_commons/validation/schemas/reana_analysis_schema.json:252]: > and also avoid that CERN users mistakenly write "htcondor" instead of "htcondorcern"

Yes, exactly, I think we should avoid this so let's remove them as currently we only support `kubernetes`, `htcondorcern` and `slurmcern`. When we will add support for other Slurm/HTCondor instances we will update the schema accordingly!

**Review by mdonadoni — APPROVED** [review]
